RNG/RandomBag{T}.cs

```diff
@@ -0,0 +1,79 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace Castle_Fayyt.RNG
+{
+  /// <summary>
+  /// Random generator of type T.
+  /// </summary>
+  /// <typeparam name="T">Type to randomly select.</typeparam>
+  internal class RandomBag<T>
+  {
+    /// <summary>
+    /// Collection of entries.
+    /// </summary>
+    private readonly List<RandomBagEntry<T>> entries = new List<RandomBagEntry<T>>();
+
+    /// <summary>
+    /// Random number generator.
+    /// </summary>
+    private readonly Random rand = new Random();
+
+    /// <summary>
+    /// Accumulated weight.
+    /// </summary>
+    private float accumulatedWeight;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="RandomBag{T}"/> class.
+    /// </summary>
+    /// <param name="entries">Collection of starting entries.</param>
+    public RandomBag(params (T item, float weight)[] entries)
+    {
+      foreach (var (item, weight) in entries)
+      {
+        this.AddEntry(item, weight);
+      }
+    }
+
+    /// <summary>
+    /// Adds entry to this random bag.
+    /// </summary>
+    /// <param name="item">Item to add to random bag.</param>
+    /// <param name="weight">Weight of item - how likely it is to be selected.</param>
+    public void AddEntry(T item, float weight)
+    {
+      this.accumulatedWeight += weight;
```

Comment:
Boom.  This is where it gets all the previous weights added to it to guarantee a full-spectrum, reasonably realistic return on weighted entries.  As a total noob who can only guess at truths half obscured by ignorance, my gut feeling is that instantiating the list bypasses this bit of code.  Yet I notice the foreach line above it which iterates over an instantiated list and adds them to an array.  So the problem might be resolved by that.  Am I getting close?

Reply:
The line `public RandomBag(params (T item, float weight)[] entries)` is called a constructor. I over-loaded the regular object constructor with this. "params" indicates that I can enter between 0 and (effectively) infinite items in the constructor which will automatically be converted into an array called entries. When the object is created, that array is passed in and a loop performed on it which adds each entry individually with `AddEntry()`.

Reply:
One thing that still bothers me about RandomBag<T>, is that choosing a random value is still O(n) complexity. I want to explore the option of using an AVL Tree to store the entries, then search for the right one. Instead of an accumulated weight, you'd have the min and max accumulated weights for correctly choosing the right node in the tree. This would reduce the selection complexity to O(log n).

Reply:
I'm going to research AVL Trees and consider.  I understand the desirability of having lower complexity, no matter the size of the program, but I also want to have control over what the program is doing.  I consider the accurate random weighting of the elements as a core dynamic of the game.  I'm loathe to choose a method with lower complexity that sacrifices my direct control on the way the battles play out.